### PR TITLE
[Feat][DeepSeek-NSA] Add fixed-length GQA sliding window forward kernel

### LINKED ISSUE #391: [FEAT][DeepSeek-NSA] add fixed-length GQA sliding window forward kernel
### Symptom / Motivation
The existing `GQAWindowSlidingKernel` (`gqa_window_sliding.py`) handles variable-length sequences via `cu_seqlens`, but has known correctness bugs (#346) and poor performance. There is no fixed-length GQA sliding window operator for prefill scenarios where all batch sequences share the same length.

### Root Cause Analysis
N/A — this is a new feature. The existing varlen kernel is a separate code path that operates differently from fixed-length prefill. A fixed-length kernel can take advantage of compile-time shape specialization and loop-range pruning to achieve better performance.

### Related Files
- `tileops/kernels/deepseek_nsa/gqa_window_sliding.py` — existing varlen kernel (reference)
- `tileops/kernels/deepseek_nsa/` — target subpackage for the new kernel
- `tileops/ops/` — Op wrapper layer

## Goal
Implement a fixed-length GQA sliding window forward kernel following FA3 window semantics:
- `k_pos <= q_pos` when `is_causal=True`
- `k_pos >= q_pos - window_size_left` when `window_size_left >= 0`
- `k_pos <= q_pos + window_size_right` when `window_size_right >= 0`

Use `window_size_left=-1` / `window_size_right=-1` to indicate no restriction on that side.

## Plan

1. Add `GqaSlidingWindowFwdKernel` (simple, sm80/89/90) in `tileops/kernels/deepseek_nsa/gqa_sliding_window_fwd.py`
   - 4 compile-time mask code paths baked into kernel via closure over `is_causal`/`window_size_left`/`window_size_right`
   - Runtime loop-range pruning (`k_start`/`k_end`) to skip masked-out KV blocks
   - NaN guard: clamp `scores_max` to `-1e38` for fully-masked KV blocks
2. Add `GqaSlidingWindowFwdWgmmaPipelinedKernel` (WGMMA pipelined, sm90) using `@T.macro` mma0/mma1
3. Export both kernels from `tileops/kernels/deepseek_nsa/__init__.py`
4. Add `GqaSlidingWindowFwdOp` wrapper in `tileops/ops/gqa_sliding_window_fwd.py`
   - Dispatches to simple kernel (sm80/89/90)
   - `total_flops` / `total_memory` properties for benchmarking
5. Add `tests/ops/test_gqa_sliding_window_fwd.py` with 14 configs covering:
   - float16 / bfloat16
   - GQA ratios 1:1, 4:1, 16:1
   - seq = 128, 256, 384, 512, 768, 2048
   - causal / non-causal, left-only / right-only / bidirectional windows, `wl=0` boundary
6. Add `benchmarks/ops/bench_gqa_sliding_window_fwd.py` with FA3 baseline comparison and warmup before profiling

## Constraints
- `window_size_left`, `window_size_right`, and `is_causal` are compile-time closure parameters baked into the kernel binary
- Fixed-length only; varlen support is a follow-up task
- Both tiers use `torch.library.custom_op` wrapper for PyTorch interop

## Acceptance Criteria
- [ ] Modified files pass unit tests
- [ ] All 14 test configs pass correctness check against pure-PyTorch reference (atol=1e-2, rtol=1e-2)
- [ ] Benchmark compiles and runs without inflated JIT latency (warmup call before profiling)
- [ ] `GqaSlidingWindowFwdOp` exported from `tileops.ops`

### PR DESCRIPTION
**Pull request by michaelwithu**
## Summary

Closes #391

Implements a fixed-length GQA sliding window forward kernel following FA3 window semantics. This is complementary to the existing varlen `GQAWindowSlidingKernel` and targets prefill scenarios where all batch sequences share the same length.

## Changes

- `tileops/kernels/deepseek_nsa/gqa_sliding_window_fwd.py`
  - `GqaSlidingWindowFwdKernel` (sm80/89/90): 4 compile-time mask code paths baked into kernel via closure, runtime loop-range pruning (`k_start`/`k_end`), NaN guard for fully-masked KV blocks
  - `GqaSlidingWindowFwdWgmmaPipelinedKernel` (sm90): WGMMA pipelined variant using `@T.macro` mma0/mma1
- `tileops/ops/gqa_sliding_window_fwd.py`: `GqaSlidingWindowFwdOp` wrapper with `total_flops`/`total_memory` properties
- `tileops/ops/__init__.py`: export `GqaSlidingWindowFwdOp`
- `tests/ops/test_gqa_sliding_window_fwd.py`: 14 correctness configs covering float16/bfloat16, GQA ratios 1:1/4:1/16:1, seq up to 2048, causal/non-causal, various window combinations
- `benchmarks/ops/bench_gqa_sliding_window_fwd.py`: benchmark vs FA3 baseline with warmup before profiling to exclude JIT compilation time

## Key Design Decisions

- `is_causal`, `window_size_left`, `window_size_right` are **compile-time closure parameters** — each unique combination compiles to a specialized kernel binary
- Fixed-length only; varlen support is a follow-up
- `GqaSlidingWindowFwdOp.forward()` returns output tensor only (not lse), consistent with `GroupQueryAttentionFwdOp`

## Benchmark Results (H200, float16)

 > Device: NVIDIA H200 | CUDA 12.8 | torch 2.9.1 | kernel: `GqaSlidingWindowFwdWgmmaPipelinedKernel` (SM90)

  | batch | seq  | heads | heads_kv | dim | is_causal | wl  | wr  | dtype    | TileOPs (ms) | TileOPs (TFLOPs/s) | FA3 (ms) | FA3 (TFLOPs/s) | Speedup |
  |-------|------|-------|----------|-----|-----------|-----|-----|----------|--------------|--------------------|----------|----------------|---------|
  | 2     | 512  | 8     | 2        | 64  | True      | -1  | -1  | float16  | 0.01         | 44.59              | 0.02     | 31.07          | 1.44×   |
  | 2     | 512  | 8     | 2        | 64  | True      | 128 | -1  | float16  | 0.01         | 28.02              | 0.02     | 12.98          | 2.16×   |
  | 2     | 512  | 8     | 2        | 64  | False     | -1  | -1  | float16  | 0.01         | 94.51              | 0.02     | 63.29          | 1.49×   |
  | 2     | 512  | 8     | 2        | 64  | False     | 64  | 64  | float16  | 0.01         | 24.91              | 0.02     | 13.81          | 1.80×   |
  | 2     | 128  | 8     | 1        | 128 | True      | 1   | -1  | float16  | 0.01         | 0.28               | 0.01     | 0.17           | 1.65×   |
  | 2     | 512  | 8     | 2        | 64  | True      | -1  | -1  | bfloat16 | 0.01         | 43.66              | 0.02     | 31.13          | 1.40×   |
  | 2     | 512  | 8     | 2        | 64  | False     | 64  | 64  | bfloat16 | 0.01         | 24.53              | 0.02     | 13.87          | 1.77×   |
  | 2     | 512  | 8     | 8        | 64  | True      | -1  | -1  | float16  | 0.01         | 43.66              | 0.02     | 30.35          | 1.44×   |
  | 2     | 512  | 16    | 1        | 64  | True      | -1  | -1  | float16  | 0.01         | 84.25              | 0.02     | 61.26          | 1.38×   |
  | 2     | 384  | 8     | 2        | 64  | True      | -1  | -1  | float16  | 0.01         | 30.13              | 0.02     | 18.17          | 1.66×   |
  | 2     | 768  | 8     | 2        | 64  | False     | 256 | -1  | float16  | 0.02         | 105.67             | 0.04     | 52.35          | 2.02×   |
  | 1     | 2048 | 8     | 2        | 64  | True      | 512 | -1  | float16  | 0.02         | 122.32             | 0.03     | 62.51          | 1.96×   |
  | 2     | 512  | 8     | 2        | 64  | False     | -1  | 64  | float16  | 0.01         | 55.60              | 0.02     | 36.05          | 1.54×   |
  | 2     | 256  | 8     | 2        | 64  | True      | 0   | -1  | float16  | 0.01         | 0.18               | 0.01     | 0.11           | 1.64×   |

  > Speedup = TileOPs TFLOPs/s ÷ FA3 TFLOPs/s. `GqaSlidingWindowFwdWgmmaPipelinedKernel` (SM90) outperforms FA3 across all 14 configurations (**1.38×–2.16×**).
  Window-constrained causal configs benefit most (causal+wl=128: 2.16×, bidirectional+wl=256: 2.02×, causal+wl=512 seq=2048: 1.96×). Rows 5 and 14 show low absolute TFLOPs/s
  due to near-zero effective KV count (wl=1 and wl=0) — TFLOPs/s is now correctly computed with the fixed `total_flops` formula.

## Next Steps

- [ ] Varlen (variable sequence length) version: fix correctness bugs in existing `GQAWindowSlidingKernel` using fixed-length kernel as reference
- [ ] Merge/refactor `tileops/kernels/flash_attn/fwd.py` GQA forward into this unified sliding window kernel
- [ ] Backward pass (`GqaSlidingWindowBwdOp`)

## Test Plan

- [x] All 14 op-level test configs pass correctness check vs pure-PyTorch reference (atol=1e-2, rtol=1e-2)
- [x] Benchmark runs without inflated JIT latency (warmup call before profiling)
- [x] `SKIP=gitleaks pre-commit run --all-files` passes

### CODE REVIEW

File: benchmarks/ops/bench_gqa_sliding_window_fwd.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        else:
+            eff_kv = t.seq
+        return 4 * t.batch * t.heads * t.seq * eff_kv * t.dim
+
+    def calculate_memory(self) -> Optional[float]:
+        """Approximate bytes accessed: read Q/K/V, write O."""
+        t = self.test
+        elem = torch.tensor([], dtype=t.dtype).element_size()
+        return (3 * t.batch * t.seq * t.heads_kv * t.dim +
+                t.batch * t.seq * t.heads * t.dim) * elem
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply benchmarks/ops/bench_gqa_sliding_window_fwd.py]: Not adopted. The pattern of computing calculate_flops / calculate_memory directly from self.test is consistent across the entire codebase (e.g. GqaFwdBenchmark in bench_gqa.py does the same). Introducing an op dependency into BenchmarkBase would break this convention and couple the benchmark to a specific op instance. We'll keep the current approach for consistency.

File: tileops/kernels/deepseek_nsa/gqa_sliding_window_fwd.py [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+                        acc_s[i, j] = T.exp2(acc_s[i, j] * scale -
+                                             scores_max[i] * scale)
+                    T.reduce_sum(acc_s, scores_sum, dim=1)
+                    for i in T.Parallel(block_m):
+                        logsum[i] = logsum[i] * scores_scale[i] + scores_sum[i]
+                    T.copy(acc_s, acc_s_cast)
+                    for i, j in T.Parallel(block_m, dim):
+                        acc_o[i, j] *= scores_scale[i]
+                    T.gemm(acc_s_cast, v_shared, acc_o,
+                           policy=T.GemmWarpPolicy.FullRow)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/kernels/deepseek_nsa/gqa_sliding_window_fwd.py:151]: Not adopted in this PR. The two kernels (GqaSlidingWindowFwdKernel and GqaSlidingWindowFwdWgmmaPipelinedKernel) have fundamentally different architectures and will be tuned independently going forward. Forcing shared macros now would increase coupling prematurely. This is a reasonable idea for a future refactoring once both kernels are stable.

File: tileops/ops/gqa_sliding_window_fwd.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            is_causal=is_causal,
+            window_size_left=window_size_left,
+            window_size_right=window_size_right,
+            dtype=dtype,
+            tune=tune,
+        )
+
+    @property
+    def default_kernel_map(self) -> Dict[str, Kernel]:
+        return {"gqa_sliding_window_fwd": GqaSlidingWindowFwdKernel}
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/ops/gqa_sliding_window_fwd.py]: Adopted and fixed. default_kernel_map now uses is_hopper() to select GqaSlidingWindowFwdWgmmaPipelinedKernel on Hopper (SM90), consistent with GroupQueryAttentionFwdOp. All 14 tests pass on H200 with the WGMMA kernel active.

File: tileops/ops/gqa_sliding_window_fwd.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                f"v shape {v.shape} does not match expected "
+                f"({self.batch}, {self.seq_len}, {self.heads_kv}, {self.dim})")
+
+        output, _ = self.kernel.forward(q, k, v)
+        return output
+
+    @property
+    def total_flops(self) -> int:
+        """Approximate FLOPs for QK^T and PV GEMMs."""
+        if self.is_causal:
```
Comment (RMLYC) [inline tileops/ops/gqa_sliding_window_fwd.py]: `total_flops` looks incorrect for `causal + window` cases. The current branch takes the `is_causal` path first and always approximates `eff_kv` as `seq_len // 2`, which ignores `window_size_left/right`. For example, with `is_causal=True` and `wl=0`, each query can only attend to `k_pos == q_pos`, so the effective KV count is ~1, not ~S/2. This will overstate TFLOPs/s in the benchmarked windowed-causal configs.

File: tileops/ops/gqa_sliding_window_fwd.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            wr = self.window_size_right if self.window_size_right >= 0 else self.seq_len
+            eff_kv = min(self.seq_len, wl + wr + 1)
+        else:
+            eff_kv = self.seq_len
+        return 4 * self.batch * self.heads * self.seq_len * eff_kv * self.dim
+
+    @property
+    def total_memory(self) -> int:
+        """Approximate bytes accessed: read Q/K/V, write O."""
+        elem = torch.tensor([], dtype=self.dtype).element_size()
```
Comment (RMLYC) [inline tileops/ops/gqa_sliding_window_fwd.py:151]: `total_memory` seems to undercount reads. This formula treats Q/K/V as if they all had `heads_kv` heads (`3 * ... * heads_kv`), but Q is shaped with `heads`, not `heads_kv`. For GQA cases where `heads > heads_kv`, that will systematically underestimate memory traffic and distort bandwidth-oriented benchmark metrics.

File: tileops/ops/gqa_sliding_window_fwd.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            Output tensor, shape [batch, seq_len, heads, dim].
+        """
+        if not q.is_contiguous():
+            q = q.contiguous()
+        if not k.is_contiguous():
+            k = k.contiguous()
+        if not v.is_contiguous():
+            v = v.contiguous()
+
+        if q.shape != (self.batch, self.seq_len, self.heads, self.dim):
```
Comment (RMLYC) [inline tileops/ops/gqa_sliding_window_fwd.py:119]: The issue asks for input validation in the op wrapper, but the current implementation only validates tensor shapes. It would be safer to also fail early on dtype/device mismatches (for example, constructing the op with `float16` and passing `bfloat16`, or passing CPU tensors), and to validate that window sizes are either `-1` or non-negative. Otherwise these become lower-level kernel/runtime errors instead of clear wrapper-level validation errors.

File: tileops/ops/gqa_sliding_window_fwd.py (hunk 5 above)
Comment (michaelwithu) [reply tileops/ops/gqa_sliding_window_fwd.py:151]: Good catch. The formula was treating Q as heads_kv-shaped. Fixed in 52e0a02: changed to 2 * (heads + heads_kv), correctly accounting for Q (read, heads), K (read, heads_kv), V (read, heads_kv), and O  (write, heads). A regression test (test_total_memory_gqa) has been added.

File: tileops/ops/gqa_sliding_window_fwd.py (hunk 6 above)
Comment (michaelwithu) [reply tileops/ops/gqa_sliding_window_fwd.py:119]: Fixed in e1ab8ec. __init__ now raises ValueError for window_size_left/right values other than -1 or ≥ 0; forward raises  ValueError early for dtype mismatches and non-CUDA tensors before reaching the kernel. Tests added in TestGqaSlidingWindowFwdOpValidation.

File: tileops/ops/gqa_sliding_window_fwd.py (hunk 4 above)
Comment (michaelwithu) [reply tileops/ops/gqa_sliding_window_fwd.py]: Fixed in 52e0a02. Replaced the if is_causal / elif window branching (which silently ignored wl/wr when causal) with a per-query loop that combines both constraints correctly. Regression tests added for wl=0 (eff_kv=1) and finite wl cases.

File: tileops/kernels/deepseek_nsa/gqa_sliding_window_fwd.py [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+def _(batch, heads, heads_kv, seq_len, dim, is_causal, window_size_left,
+      window_size_right, dtype, block_m, block_n, num_stages, threads,
+      *inputs):
+    fake_o = torch.empty_like(inputs[0])
+    fake_lse = fake_o.new_empty([batch, heads, seq_len])
+    return fake_o, fake_lse
+
+
+class GqaSlidingWindowFwdKernel(Kernel):
+    supported_archs: list[int] = [80, 89, 90]
```
Comment (lcy-seso) [inline tileops/kernels/deepseek_nsa/gqa_sliding_window_fwd.py:189]: Have architectures 90, 89, and 90 all been tested? If not, please remove any untested architectures.

**Review by lcy-seso — APPROVED** [review]
